File: src/parsers/claude.ts
/**
 * parsers/claude.ts — Claude Code sessions.
 *
 * Store: one JSONL per session at ~/.claude/projects/<enc-cwd>/<uuid>.jsonl.
 * Event stream dispatched on top-level `type`. Ground-truthed 2026-07.
 * Relevant record types:
 *   user / assistant       chat turns; message.content is str (user) or block list
 *   system                 subtype ∈ compact_boundary | local_command | away_summary |
 *                          informational | turn_duration | stop_hook_summary |
 *                          api_error | scheduled_task_fire | agents_killed → system event
 *   ai-title               carries aiTitle → native session title
 *   mode / permission-mode → system event (mode/permission change)
 *   attachment             → system event
 *   file-history-snapshot / last-prompt → skipped (pure engine bookkeeping)
 *
 * One assistant record's content list can mix text, thinking, and tool_use
 * blocks — we emit ONE EVENT PER BLOCK. A user record carrying toolUseResult is
 * a tool result (blocks are tool_result payloads, linked by tool_use_id).
 * cwd/gitBranch/sessionId/uuid/parentUuid come off records directly. Tree A =
 * parentUuid→uuid; tree B = forkedFrom{sessionId,messageUuid} (fork/resume) and
 * subagents.
 *
 * SUBAGENTS (tree B, resolved deterministically here — NOT deferred to fan-out).
 * A subagent transcript is a nested `bg`/isSidechain file at
 *   <enc-cwd>/<parent-uuid>/subagents/agent-<hash>.jsonl
 * with a sibling `agent-<hash>.meta.json` carrying `{agentType}`. The parent is
 * the DIRECTORY NAME, so parent→child needs no LLM. The subagent's own
 * `sessionId` field equals the parent uuid (would collide), so identity is the
 * PATH: id = `claude:<parent-uuid>/agent-<hash>`, parentSessionId =
 * `claude:<parent-uuid>`. That edge feeds the existing tree-B lineage collapse.
 */

import { readdirSync, statSync } from "fs";
import { join } from "path";
import { CLAUDE_PROJECTS } from "../paths.js";
import type { Event, Parser, UnifiedSession } from "../types.js";
import {
  asDict,
  asNum,
  asStr,
  compact,
  EventBuilder,
  fileSource,
  flattenContent,
  isoToEpochSec,
  readJson,
  readJsonl,
} from "../util.js";

const UUID_RE = /^[0-9a-fA-F-]{36}$/;

/** A discovered transcript: a top-level session file or a nested subagent file. */
interface FileEntry {
  path: string;
  /** For a subagent file: the parent session's native uuid (its dir name). */
  parentUuid: string | null;
}

/**
 * Enumerate Claude transcripts. Two shapes under each project dir:
 *   <proj>/<uuid>.jsonl                                  → top-level session
 *   <proj>/<uuid>/subagents/agent-<hash>.jsonl           → subagent (tree B)
 * The top-level scan is one level deep (unchanged); the subagent scan descends
 * into each `<uuid>/subagents/` dir. Non-subagent subdirs are ignored.
 */
function listFiles(): FileEntry[] {
  let projectDirs: string[];
  try {
    projectDirs = readdirSync(CLAUDE_PROJECTS);
  } catch {
    return [];
  }
  const files: FileEntry[] = [];
  for (const proj of projectDirs) {
    const dir = join(CLAUDE_PROJECTS, proj);
    let entries: string[];
    try {
      if (!statSync(dir).isDirectory()) continue;
      entries = readdirSync(dir);
    } catch {
      continue;
    }
    for (const e of entries) {
      if (e.endsWith(".jsonl")) {
        files.push({ path: join(dir, e), parentUuid: null });
        continue;
      }
      // <uuid>/subagents/agent-<hash>.jsonl — nested subagent transcripts
      const subDir = join(dir, e, "subagents");
      let subEntries: string[];
      try {
        if (!statSync(subDir).isDirectory()) continue;
        subEntries = readdirSync(subDir);
      } catch {
        continue;
      }
      for (const se of subEntries)
        if (se.endsWith(".jsonl")) files.push({ path: join(subDir, se), parentUuid: e });
    }
  }
  return files;
}

interface Scan {
  sessionId: string | null;
  cwd: string | null;
  gitBranch: string | null;
  title: string | null;
  model: string | null;
  forkedFrom: string | null;
  version: string | null;
  sessionKind: string | null;
  startedAt: number | null;
  endedAt: number | null;
  records: number;
  turns: number;
}

/** Single pass over a transcript for listing metadata + fingerprint inputs. */
function scan(path: string): Scan {
  const s: Scan = {
    sessionId: null,
    cwd: null,
    gitBranch: null,
    title: null,
    model: null,
    forkedFrom: null,
    version: null,
    sessionKind: null,
    startedAt: null,
    endedAt: null,
    records: 0,
    turns: 0,
  };
  for (const r of readJsonl(path)) {
    s.records++;
    s.sessionId ??= asStr(r.sessionId);
    s.cwd ??= asStr(r.cwd);
    s.gitBranch ??= asStr(r.gitBranch);
    s.version ??= asStr(r.version);
    s.sessionKind ??= asStr(r.sessionKind);
    const ts = isoToEpochSec(r.timestamp);
    if (ts != null) {
      s.startedAt ??= ts;
      s.endedAt = ts;
    }
    const type = asStr(r.type);
    if (type === "ai-title") s.title ??= asStr(r.aiTitle);
    if (type === "assistant") s.model ??= asStr(asDict(r.message).model);
    if (r.forkedFrom && !s.forkedFrom) {
      const f = asDict(r.forkedFrom);
      const sid = asStr(f.sessionId);
      if (sid) s.forkedFrom = `claude:${sid}#${asStr(f.messageUuid) ?? ""}`;
    }
    if (type === "user" && !r.toolUseResult) s.turns++;
    else if (type === "assistant") s.turns++;
  }
  return s;
}

/** Emit events for one assistant content block. */
function pushAssistantBlock(
  b: Record<string, unknown>,
  ts: number | null,
  nativeId: string | null,
  parentId: string | null,
  tokens: number | null,
  out: EventBuilder
): void {
  const btype = asStr(b.type) ?? "text";
  const base = { nativeId, parentId };
  if (btype === "text") {
    out.push("assistant", asStr(b.text) ?? "", ts, { ...base, tokens });
  } else if (btype === "thinking") {
    out.push("thinking", asStr(b.thinking) ?? asStr(b.text) ?? "", ts, base);
  } else if (btype === "tool_use") {
    out.push("tool_call", compact(b.input ?? b), ts, { ...base, name: asStr(b.name), callId: asStr(b.id) });
  } else if (btype === "tool_result") {
    out.push("tool_result", compact(b.content ?? b), ts, {
      ...base,
      callId: asStr(b.tool_use_id),
      isError: b.is_error === true,
    });
  } else {
    out.push("system", compact(b), ts, { ...base, subtype: btype });
  }
}

/** A tool_result block on a user record (carries tool_use_id + is_error). */
function pushToolResultBlock(
  b: Record<string, unknown>,
  ts: number | null,
  nativeId: string | null,
  parentId: string | null,
  out: EventBuilder
): void {
  out.push("tool_result", compact(b.content ?? b), ts, {
    nativeId,
    parentId,
    callId: asStr(b.tool_use_id),
    isError: b.is_error === true,
  });
}

export const claudeParser: Parser = {
  agent: "claude",

  listSessions(sinceEpochSec) {
    const out: UnifiedSession[] = [];
    for (const { path, parentUuid } of listFiles()) {
      if (sinceEpochSec != null) {
        try {
          if (statSync(path).mtimeMs / 1000 < sinceEpochSec) continue;
        } catch {
          continue;
        }
      }
      const m = scan(path);

      if (parentUuid) {
        // Subagent: identity is the PATH (its sessionId equals the parent uuid,
        // so it can't be the key). hash = filename; agentType from the sidecar.
        const hash = path.replace(/\.jsonl$/, "").split("/").pop() ?? "";
        if (!hash) continue;
        const nativeId = `${parentUuid}/${hash}`;
        const meta = readJson(path.replace(/\.jsonl$/, ".meta.json"));
        const agentType = asStr(asDict(meta).agentType);
        out.push({
          id: `claude:${nativeId}`,
          agent: "claude",
          cwd: m.cwd,
          startedAt: m.startedAt,
          endedAt: m.endedAt,
          title: agentType ? `[${agentType} subagent]` : null, // no ai-title in bg files
          model: m.model,
          provider: null,
          gitBranch: m.gitBranch,
          parentSessionId: `claude:${parentUuid}`,
          forkedFrom: m.forkedFrom,
          turnCount: m.turns,
          eventCount: m.turns, // refined on read
          sources: [fileSource(path, "jsonl")],
          fingerprint: `${m.records}:${m.endedAt ?? 0}`,
          meta: { version: m.version, sessionKind: m.sessionKind, agentType, isSubagent: true },
        });
        continue;
      }

      const nativeId = m.sessionId ?? path.replace(/\.jsonl$/, "").split("/").pop() ?? "";
      if (!m.sessionId && !UUID_RE.test(nativeId)) continue;
      out.push({
        id: `claude:${nativeId}`,
        agent: "claude",
        cwd: m.cwd,
        startedAt: m.startedAt,
        endedAt: m.endedAt,
        title: m.title,
        model: m.model,
        provider: null,
        gitBranch: m.gitBranch,
        parentSessionId: null,
        forkedFrom: m.forkedFrom,
        turnCount: m.turns,
        eventCount: m.turns, // refined on read
        sources: [fileSource(path, "jsonl")],
        fingerprint: `${m.records}:${m.endedAt ?? 0}`,
        meta: { version: m.version, sessionKind: m.sessionKind },
      });
    }
    return out;
  },

  readEvents(nativeId): Event[] {
    // Subagent id is "<parent-uuid>/agent-<hash>" → nested at
    // <…>/<parent-uuid>/subagents/agent-<hash>.jsonl.
    const slash = nativeId.indexOf("/");
    const suffix =
      slash >= 0
        ? `/${nativeId.slice(0, slash)}/subagents/${nativeId.slice(slash + 1)}.jsonl`
        : `/${nativeId}.jsonl`;

    let target: string | null = null;
    for (const { path } of listFiles()) {
      if (path.endsWith(suffix)) {
        target = path;
        break;
      }
    }
    if (!target && slash < 0) {
      // top-level fallback: some files are named by uuid but carry a different sessionId
      for (const { path, parentUuid } of listFiles()) {
        if (parentUuid) continue;
        for (const r of readJsonl(path)) {
          if (asStr(r.sessionId) === nativeId) {
            target = path;
            break;
          }
        }
        if (target) break;
      }
    }
    if (!target) return [];

    const out = new EventBuilder();
    for (const r of readJsonl(target)) {
      const type = asStr(r.type);
      const ts = isoToEpochSec(r.timestamp);
      const uuid = asStr(r.uuid);
      const parent = asStr(r.parentUuid);

      if (type === "system") {
        out.push("system", flattenContent(r.content ?? r.data), ts, {
          subtype: asStr(r.subtype),
          nativeId: uuid,
          parentId: parent,
        });
        continue;
      }
      if (type === "mode" || type === "permission-mode") {
        out.push("system", compact(r.mode ?? r.permissionMode ?? r), ts, {
          subtype: type,
          nativeId: uuid,
          parentId: parent,
        });
        continue;
      }
      if (type === "attachment") {
        out.push("system", compact(r.attachment ?? r), ts, {
          subtype: "attachment",
          nativeId: uuid,
          parentId: parent,
        });
        continue;
      }
      if (type !== "user" && type !== "assistant") continue; // last-prompt/file-history-snapshot/ai-title: engine bookkeeping

      const msg = asDict(r.message);
      const content = msg.content;

      if (type === "user") {
        if (r.toolUseResult) {
          if (Array.isArray(content)) {
            for (const b of content)
              if (b && typeof b === "object")
                pushToolResultBlock(b as Record<string, unknown>, ts, uuid, parent, out);
          } else {
            out.push("tool_result", compact(r.toolUseResult), ts, { nativeId: uuid, parentId: parent });
          }
        } else {
          out.push("user", flattenContent(content), ts, { nativeId: uuid, parentId: parent });
        }
        continue;
      }

      // assistant: one event per content block, usage attached to the text block
      const usage = asDict(msg.usage);
      const tok = (asNum(usage.input_tokens) ?? 0) + (asNum(usage.output_tokens) ?? 0) || null;
      if (Array.isArray(content)) {
        for (const b of content)
          if (b && typeof b === "object")
            pushAssistantBlock(b as Record<string, unknown>, ts, uuid, parent, tok, out);
      } else {
        out.push("assistant", flattenContent(content), ts, { nativeId: uuid, parentId: parent, tokens: tok });
      }
    }
    return out.events;
  },
};

File: src/parsers/claude.rs
use super::{Parser, SessionCandidate};
use crate::model::{
    compact, flatten, Agent, Capture, Event, EventKind, NativeSource, ParsedSession, Session,
};
use anyhow::Result;
use chrono::DateTime;
use serde_json::{json, Value};
use std::{fs, path::Path};
use walkdir::WalkDir;

pub struct ClaudeParser;
fn s(v: Option<&Value>) -> Option<String> {
    v.and_then(Value::as_str).map(str::to_owned)
}
fn ts(v: Option<&Value>) -> Option<i64> {
    v.and_then(Value::as_str)
        .and_then(|x| DateTime::parse_from_rfc3339(x).ok())
        .map(|x| x.timestamp_millis())
}
fn ev(k: EventKind, text: String, r: &Value, t: i64) -> Event {
    let mut e = Event::new(k, text);
    e.native_id = s(r.get("uuid"));
    e.parent_id = s(r.get("parentUuid"));
    e.created_at_ms = Some(t);
    e
}
fn rel(root: &Path, p: &Path) -> String {
    p.strip_prefix(root)
        .ok()
        .and_then(|x| x.to_str())
        .unwrap_or_else(|| {
            p.file_name()
                .and_then(|x| x.to_str())
                .unwrap_or("session.jsonl")
        })
        .replace(std::path::MAIN_SEPARATOR, "/")
}

fn parse(path: &Path, root: &Path, candidate: &SessionCandidate) -> Result<ParsedSession> {
    let text = fs::read_to_string(path)?;
    let records: Vec<Value> = text
        .lines()
        .filter_map(|l| serde_json::from_str(l).ok())
        .collect();
    let mut id = None;
    let mut cwd = None;
    let mut branch = None;
    let mut title = None;
    let mut model = None;
    let mut started = None;
    let mut ended = None;
    let mut forked = None;
    let mut events = Vec::new();
    for r in &records {
        let t = ts(r.get("timestamp"));
        if started.is_none() {
            started = t;
        }
        if t.is_some() {
            ended = t;
        }
        id = id.or_else(|| s(r.get("sessionId")));
        if forked.is_none() {
            forked = r.get("forkedFrom").and_then(|value| {
                let session_id = s(value.get("sessionId"))?;
                let message_uuid = s(value.get("messageUuid"));
                Some(match message_uuid {
                    Some(message_uuid) => format!("claude:{session_id}#{message_uuid}"),
                    None => format!("claude:{session_id}"),
                })
            });
        }
        cwd = cwd.or_else(|| s(r.get("cwd")));
        branch = branch.or_else(|| s(r.get("gitBranch")));
        if r.get("type").and_then(Value::as_str) == Some("ai-title") {
            title = title.or_else(|| s(r.get("aiTitle")));
        }
        if r.get("type").and_then(Value::as_str) == Some("assistant") {
            model = model.or_else(|| r.get("message").and_then(|m| s(m.get("model"))));
        }
        match r.get("type").and_then(Value::as_str).unwrap_or("") {
            "user" => {
                if r.get("toolUseResult").is_some() {
                    let mut e = ev(
                        EventKind::ToolResult,
                        compact(r.get("toolUseResult").unwrap()),
                        r,
                        t.unwrap_or_default(),
                    );
                    e.subtype = Some("tool_result".into());
                    events.push(e)
                } else {
                    events.push(ev(
                        EventKind::User,
                        flatten(r.get("message").unwrap_or(&Value::Null)),
                        r,
                        t.unwrap_or_default(),
                    ));
                }
            }
            "assistant" => {
                let c = r
                    .get("message")
                    .and_then(|m| m.get("content"))
                    .unwrap_or(&Value::Null);
                if let Some(arr) = c.as_array() {
                    for b in arr {
                        let typ = b.get("type").and_then(Value::as_str).unwrap_or("text");
                        let (k, txt) = match typ {
                            "thinking" => (
                                EventKind::Thinking,
                                s(b.get("thinking").or_else(|| b.get("text"))),
                            ),
                            "tool_use" => (
                                EventKind::ToolCall,
                                Some(compact(b.get("input").unwrap_or(b))),
                            ),
                            "tool_result" => (
                                EventKind::ToolResult,
                                Some(compact(b.get("content").unwrap_or(b))),
                            ),
                            _ => (EventKind::Assistant, s(b.get("text"))),
                        };
                        let mut e = ev(k, txt.unwrap_or_default(), r, t.unwrap_or_default());
                        e.subtype = Some(typ.into());
                        e.name = s(b.get("name"));
                        e.call_id = s(b.get("id").or_else(|| b.get("tool_use_id")));
                        e.is_error = b.get("is_error").and_then(Value::as_bool);
                        events.push(e)
                    }
                } else {
                    events.push(ev(
                        EventKind::Assistant,
                        flatten(c),
                        r,
                        t.unwrap_or_default(),
                    ));
                }
            }
            "system" | "mode" | "permission-mode" | "attachment" => {
                let mut e = ev(
                    EventKind::System,
                    compact(
                        r.get("content")
                            .or_else(|| r.get("attachment"))
                            .unwrap_or(r),
                    ),
                    r,
                    t.unwrap_or_default(),
                );
                e.subtype = r
                    .get("type")
                    .and_then(Value::as_str)
                    .map(str::to_owned)
                    .or_else(|| s(r.get("subtype")));
                events.push(e)
            }
            _ => {}
        }
    }
    let id = id.unwrap_or_else(|| {
        path.file_stem()
            .and_then(|x| x.to_str())
            .unwrap_or("unknown")
            .into()
    });
    let mut sources = vec![NativeSource {
        locator: path.display().to_string(),
        kind: "jsonl".into(),
        restore_path: rel(root, path),
        role: None,
        bytes: candidate.bytes,
        mtime_ns: candidate.mtime_ns,
        mode: None,
        capture: Some(Capture::File {
            path: path.display().to_string(),
        }),
    }];
    let sidecar = path.with_extension("meta.json");
    if sidecar.exists() {
        if let Ok(sm) = fs::metadata(&sidecar) {
            sources.push(NativeSource {
                locator: sidecar.display().to_string(),
                kind: "json".into(),
                restore_path: rel(root, &sidecar),
                role: Some("subagent-meta".into()),
                bytes: Some(sm.len() as i64),
                mtime_ns: None,
                mode: None,
                capture: Some(Capture::File {
                    path: sidecar.display().to_string(),
                }),
            });
        }
    }
    Ok(ParsedSession {
        session: Session {
            id: format!("claude:{id}"),
            agent: Agent::Claude,
            cwd,
            started_at_ms: started,
            ended_at_ms: ended,
            title,
            model,
            provider: None,
            git_branch: branch,
            parent_session_id: None,
            forked_from: forked,
            meta: json!({"recordCount":records.len()}),
            fingerprint: format!("{}:{}", records.len(), ended.unwrap_or_default()),
            sources,
        },
        events,
    })
}
impl Parser for ClaudeParser {
    fn agent(&self) -> Agent {
        Agent::Claude
    }
    fn discover(&self, root: &Path) -> Result<Vec<SessionCandidate>> {
        let mut out = Vec::new();
        if !root.exists() {
            return Ok(out);
        };
        for e in WalkDir::new(root)
            .follow_links(false)
            .into_iter()
            .filter_map(Result::ok)
        {
            if e.file_type().is_file() && e.path().extension().is_some_and(|x| x == "jsonl") {
                if let Ok(mut candidate) = SessionCandidate::file(e.path().to_path_buf()) {
                    candidate.include_file(&e.path().with_extension("meta.json"));
                    out.push(candidate);
                }
            }
        }
        Ok(out)
    }

    fn parse(&self, candidate: &SessionCandidate, root: &Path) -> Result<Option<ParsedSession>> {
        let mut parsed = parse(&candidate.path, root, candidate)?;
        let rel_path = candidate.path.strip_prefix(root).unwrap_or(&candidate.path);
        let parts: Vec<_> = rel_path.components().collect();
        if let Some(index) = parts
            .iter()
            .position(|part| part.as_os_str() == "subagents")
        {
            if index > 0 {
                let parent = parts[index - 1].as_os_str().to_string_lossy();
                let child = candidate
                    .path
                    .file_stem()
                    .and_then(|value| value.to_str())
                    .unwrap_or("agent");
                parsed.session.id = format!("claude:{parent}/{child}");
                parsed.session.parent_session_id = Some(format!("claude:{parent}"));
                let meta_path = candidate.path.with_extension("meta.json");
                if let Ok(meta_text) = fs::read_to_string(meta_path) {
                    if let Ok(meta) = serde_json::from_str::<Value>(&meta_text) {
                        parsed.session.meta["agentType"] =
                            meta.get("agentType").cloned().unwrap_or(Value::Null);
                        parsed.session.meta["isSubagent"] = Value::Bool(true);
                    }
                }
            }
        }
        Ok(Some(parsed))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn captures_forked_from_session_and_message() {
        let dir = tempdir().unwrap();
        let path = dir.path().join("fork.jsonl");
        fs::write(
            &path,
            [json!({
                "sessionId":"child",
                "forkedFrom":{"sessionId":"parent","messageUuid":"m-1"},
                "type":"user",
                "timestamp":"2026-08-19T00:00:00Z",
                "message":{"role":"user","content":"continue"}
            })
            .to_string()]
            .join("\n"),
        )
        .unwrap();
        let parsed = ClaudeParser.parse_all(dir.path()).unwrap();
        assert_eq!(
            parsed[0].session.forked_from.as_deref(),
            Some("claude:parent#m-1")
        );
    }
}
